allennlp/tests/modules/sampled_softmax_loss_test.py

```diff
@@ -0,0 +1,38 @@
+# pylint: disable=no-self-use,invalid-name,protected-access
+import torch
+
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.modules.sampled_softmax_loss import _choice, SampledSoftmaxLoss
+from allennlp.models.bidirectional_lm import _SoftmaxLoss
+
+
+class TestSampledSoftmaxLoss(AllenNlpTestCase):
+    def test_choice(self):
+        sample, num_samples = _choice(num_words=1000, num_samples=50)
+        assert len(set(sample)) == 50
+        assert all(0 <= x < 1000 for x in sample)
+        assert num_samples >= 50
+
+    def test_sampled_softmax_can_run(self):
+        softmax = SampledSoftmaxLoss(num_words=1000, embedding_dim=12, num_samples=50)
+
+        # sequence_length, embedding_dim
+        embedding = torch.rand(100, 12)
+        targets = torch.randint(0, 1000, (100,)).long()
+
+        _ = softmax(embedding, targets)
+
+    def test_sampled_almost_equals_unsampled_when_num_samples_is_almost_all(self):
+        sampled_softmax = SampledSoftmaxLoss(num_words=10000, embedding_dim=12, num_samples=9999)
+        unsampled_softmax = _SoftmaxLoss(num_words=10000, embedding_dim=12)
+
+        # sequence_length, embedding_dim
+        embedding = torch.rand(100, 12)
+        targets = torch.randint(0, 1000, (100,)).long()
+
+        full_loss = unsampled_softmax(embedding, targets).item()
+        sampled_loss = sampled_softmax(embedding, targets).item()
+
+        # Should be really close
+        pct_error = (sampled_loss - full_loss) / full_loss
+        assert abs(pct_error) < 0.01
```

Comment:
Just looking at this PR now, this test actually shouldn't pass because the `SampledSoftmaxLoss` and the `_SoftmaxLoss` have different parameters!  I'm guessing the < 0.01 percent error is far too loose of a check, the values should be much closer.

To implement this test, it's necessary to set manually set the weights on one of the softmax instances to match the weights on the other after initializing.

Then we should observe:

- in `eval` mode, the loss should be exactly the same for SampledSoftmax and Softmax (to near machine precision, e.g. absolute difference < 1e-6 for float32)
- the loss for SampledSoftmax in eval mode is > SampledSoftmax in train mode

To check SampledSoftmax in train mode, I'd recommend stubbing out the sampler (`choice` function) with something deterministic that returns a list of known ids.  Then set all of the bias parameters for the ids not in this list of something very small (-10000) so their contribution to the full softmax term is zero.  Then, using the deterministic sampler the loss in `train` mode should be the same as loss in `eval` mode.

Reply:
Right! Thanks for catching this, @matt-peters.